constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,534 @@
+    # hasher[0], ..., hasher[7] are absorbed into the hasher.
+    let u_span = f_span(op_bits) * h_init
+
+    # When RESPAN operation is executed, contents of hasher[0], ..., hasher[7] (which contain the 
+    # new operation batch) are absorbed into the hasher.
+    let u_respan = f_respan(op_bits, extra) * h_abp
+
+    # When END operation is executed, the hash result is copied into registers 
+    # hasher[0], ..., hasher[3].
+    let u_end = f_end(op_bits, extra) * h_res
+
+    # Enforce the block hash computation constraint.
+    enf p[0]' * (u_ctrli + u_span + u_respan + u_end + 1 - (f_ctrli(op_bits, extra) + 
+        f_span(op_bits) + f_respan(op_bits, extra) + f_end(op_bits, extra))) = p[0]
+
+
+# Enforces constraint for updating the block stack table.
+#
+# Constraint degree: 8
+ev block_stack_table_constraints([addr, op_bits[7], hasher[8], s0], [extra, p[4]]):
+    # When JOIN operation is executed, row (addr', addr, 0) is added to the block stack table.
+    let v_join = f_join(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When SPLIT operation is executed, row (addr', addr, 0) added to the block stack table.
+    let v_split = f_split(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When LOOP operation is executed, row (addr', addr, 1) is added to the block stack table if 
+    # the value at the top of the operand stack is 1, and row (addr', addr, 0) is added to the 
+    # block stack table if the value at the top of the operand stack is 0.
+    let v_loop = f_loop(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr + alpha[3] * s0)
+
+    # When SPAN operation is executed, row (addr', addr, 0) is added to the block stack table.
+    let v_span = f_span(op_bits) * (alpha[0] + alpha[1] * addr' + alpha[2] * addr)
+
+    # When RESPAN operation is executed, row (addr, hasher[1]', 0) is removed from the block stack 
+    # table, and row (addr', hasher[1]', 0) is added to the table. The prover sets the value of 
+    # register hasher[1] at the next row to the ID of the parent block:
+    let u_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr + alpha[2] * hasher[1]')
+    let v_respan = f_respan(op_bits, extra) * (alpha[0] + alpha[1] * addr' + alpha[2] * hasher[1]')
+
+    # When END operation is executed, row (addr, addr', hasher[5]) is removed from the block span 
+    # table. Register hasher[5] contains the is_loop flag.
+    let u_end = f_end(op_bits, extra) * 
+        (alpha[0] + alpha[1] * addr + alpha[2] * addr' + alpha[3] * hasher[5])
+
+    # Enforce the block stack table constraint.
+    # TODO: do proper line wrapping
+    enf p[1]' * (u_end + u_respan + 1 - (f_end(op_bits, extra) + f_respan(op_bits, extra)))
+        = p[1] * (v_join + v_split + v_loop + v_span + v_respan + 1 - (f_join(op_bits) + 
+        f_split(op_bits) + f_loop(op_bits) + f_span(op_bits) + 
+        f_respan(op_bits, extra)))
```

Comment:
A few comments:
1. let's define the flags as variables above the relevant `v_*` variables above instead of recomputing them
2. For line wrapping, I would put the `=` on the first line and then put everything after the `-` on its own line. (or maybe you can even fit everything after the `=` on one line, once you make the change in (1). Then I think you can remove the `TODO`

This should combine to give something readable like the following:
```
enf p[1]' * (u_end + u_respan + 1 - (f_end + f_respan)) =
        p[1] * (v_join + v_split + v_loop + v_span + v_respan + 1 - 
               (f_split + f_loop + f_span + f_respan))
```